app/scripts/controllers/network/fake-provider.ts

```diff
@@ -0,0 +1,224 @@
+   * of specific invocations of `sendAsync` matching a `method`.
+   */
+  constructor({ stubs = [] }: FakeProviderEngineOptions) {
+    super({ engine: new JsonRpcEngine() });
+    this.#originalStubs = stubs;
+    this.#stubs = this.#originalStubs.slice();
+    this.calledStubs = [];
+  }
+
+  send = <Params, Response>(
+    payload: JsonRpcRequest<Params>,
+    callback: (error: unknown, response?: JsonRpcResponse<Response>) => void,
+  ) => {
+    return this.#handleSend(payload, callback);
+  };
+
+  sendAsync = <Params, Response>(
+    payload: JsonRpcRequest<Params>,
+    callback: (error: unknown, response?: JsonRpcResponse<Response>) => void,
+  ) => {
+    return this.#handleSend(payload, callback);
+  };
+
+  #handleSend<Params, Response>(
+    payload: JsonRpcRequest<Params>,
+    callback?: (error: unknown, response?: JsonRpcResponse<Response>) => void,
+  ) {
+    if (Array.isArray(payload)) {
+      throw new Error("Arrays aren't supported");
+    }
+
+    const index = this.#stubs.findIndex((stub) => {
+      return (
+        stub.request.method === payload.method &&
+        (!('params' in stub.request) ||
+          isDeepStrictEqual(stub.request.params, payload.params))
+      );
+    });
+
+    if (index === -1) {
+      const matchingCalledStubs = this.calledStubs.filter((stub) => {
+        return (
+          stub.request.method === payload.method &&
+          (!('params' in stub.request) ||
+            isDeepStrictEqual(stub.request.params, payload.params))
+        );
+      });
+      let message = `Could not find any stubs matching: ${inspect(payload, {
+        depth: null,
+      })}`;
+      if (matchingCalledStubs.length > 0) {
+        message += `\n\nIt appears the following stubs were defined, but have been called already:\n\n${inspect(
+          matchingCalledStubs,
+          { depth: null },
+        )}`;
+      }
+
+      throw new Error(message);
+    } else {
+      const stub = this.#stubs[index];
+
+      if (stub.discardAfterMatching !== false) {
+        this.#stubs.splice(index, 1);
+      }
+
+      if (callback === undefined) {
+        this.#handleRequest(stub, (error, response) => {
+          console.warn(
+            'got result or error but not doing anything with it',
+            error,
+            response,
+          );
+        });
+      } else if (stub.delay) {
+        originalSetTimeout(() => {
+          this.#handleRequest(stub, callback);
+        }, stub.delay);
+      } else {
+        this.#handleRequest(stub, callback);
+      }
+
+      this.calledStubs.push({ ...stub });
+    }
+  }
+
+  #handleRequest<Params, Response>(
+    stub: FakeProviderStub<Params, Response>,
+    callback: (error: unknown, response?: JsonRpcResponse<Response>) => void,
+  ) {
+    const promiseForBeforeCompleting =
+      stub.beforeCompleting === undefined
+        ? Promise.resolve()
+        : Promise.resolve(stub.beforeCompleting());
+
+    promiseForBeforeCompleting.then(() => {
```

Comment:
Nit: Could we use async/await here rather than `.then`? Otherwise any stack traces in tests will terminate here, rather than showing further up the stack

Reply:
Good point. This method wouldn't be awaited, so as long as that's not confusing then this is fine. I've updated it in [1e5c662](https://github.com/MetaMask/metamask-extension/pull/18628/commits/1e5c662ed5c4629b8c7423d8900ce53486b64e46).